File: go/bootstrap.go
package allwright

import (
	"archive/tar"
	"archive/zip"
	"bytes"
	"compress/gzip"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"sync"
	"time"

	enginev1 "allwright.dev/gen/allwright/engine/v1"

	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

const (
	allwrightAutoInstallEnvVar = "ALLWRIGHT_AUTO_INSTALL"
	allwrightCLIPathEnvVar     = "ALLWRIGHT_CLI_PATH"
	allwrightHomeEnvVar        = "ALLWRIGHT_HOME"
	allwrightRepositoryEnvVar  = "ALLWRIGHT_REPOSITORY"
	allwrightVersionEnvVar     = "ALLWRIGHT_VERSION"
	defaultReleaseRepository   = "allwright-dev/allwright"
	defaultReleaseVersion      = "0.0.42"
)

var bootstrapState struct {
	mu                   sync.Mutex
	managedServerBase    string
	managedServerAddr    string
	managedServer        *exec.Cmd
	managedServerStdout  *tailBuffer
	managedServerStderr  *tailBuffer
	managedServerExited  bool
	managedServerExitErr error
}

type pingStatus struct {
	version string
}

type tailBuffer struct {
	mu   sync.Mutex
	data []byte
}

func (b *tailBuffer) Write(p []byte) (int, error) {
	b.mu.Lock()
	defer b.mu.Unlock()

	const maxBytes = 8_000
	if len(p) >= maxBytes {
		b.data = append([]byte(nil), p[len(p)-maxBytes:]...)
		return len(p), nil
	}

	combined := append(b.data, p...)
	if len(combined) > maxBytes {
		combined = combined[len(combined)-maxBytes:]
	}
	b.data = append([]byte(nil), combined...)
	return len(p), nil
}

func (b *tailBuffer) String() string {
	b.mu.Lock()
	defer b.mu.Unlock()
	return strings.TrimSpace(string(b.data))
}

func ensureRuntimeReady(ctx context.Context, serverAddr string) (string, error) {
	expectedVersion := expectedRuntimeVersion()
	if status, err := pingServer(ctx, serverAddr); err == nil {
		if status.version == expectedVersion {
			return serverAddr, nil
		}
		if !isLocalServerAddr(serverAddr) {
			return "", fmt.Errorf(
				"allwright server at %s is running version %s but this client expects %s",
				serverAddr,
				displayVersion(status.version),
				expectedVersion,
			)
		}
	}
	if !isLocalServerAddr(serverAddr) {
		return "", fmt.Errorf("allwright could not reach engine server at %s; automatic startup is only supported for local addresses", serverAddr)
	}

	var managedServerAddr string
	bootstrapState.mu.Lock()
	if bootstrapState.managedServer != nil && bootstrapState.managedServer.ProcessState == nil && bootstrapState.managedServerBase == serverAddr {
		managedServerAddr = bootstrapState.managedServerAddr
	}
	if managedServerAddr == "" && bootstrapState.managedServer != nil && bootstrapState.managedServer.Process != nil {
		_ = bootstrapState.managedServer.Process.Kill()
		bootstrapState.managedServer = nil
		bootstrapState.managedServerAddr = ""
		bootstrapState.managedServerBase = ""
		bootstrapState.managedServerStdout = nil
		bootstrapState.managedServerStderr = nil
		bootstrapState.managedServerExited = false
		bootstrapState.managedServerExitErr = nil
	}
	bootstrapState.mu.Unlock()

	if managedServerAddr != "" {
		return waitForServer(ctx, managedServerAddr, expectedVersion)
	}

	cliPath, err := ensureCLIAvailable(expectedVersion)
	if err != nil {
		return "", err
	}
	if err := ensureWebPlugin(cliPath, expectedVersion); err != nil {
		return "", err
	}

	resolvedServerAddr := serverAddr
	if status, err := pingServer(ctx, serverAddr); err == nil && status.version != expectedVersion {
		resolvedServerAddr, err = allocateManagedServerAddr(serverAddr)
		if err != nil {
			return "", err
		}
	}

	cmd := exec.Command(cliPath, "serve", "--listen-addr", cliListenAddr(resolvedServerAddr))
	cmd.Stdin = nil
	stdoutBuf := &tailBuffer{}
	stderrBuf := &tailBuffer{}
	cmd.Stdout = stdoutBuf
	cmd.Stderr = stderrBuf
	if err := cmd.Start(); err != nil {
		return "", fmt.Errorf("start allwright server with %s: %w", cliPath, err)
	}

	bootstrapState.mu.Lock()
	bootstrapState.managedServerBase = serverAddr
	bootstrapState.managedServer = cmd
	bootstrapState.managedServerAddr = resolvedServerAddr
	bootstrapState.managedServerStdout = stdoutBuf
	bootstrapState.managedServerStderr = stderrBuf
	bootstrapState.managedServerExited = false
	bootstrapState.managedServerExitErr = nil
	bootstrapState.mu.Unlock()

	go func(cmd *exec.Cmd) {
		err := cmd.Wait()
		bootstrapState.mu.Lock()
		defer bootstrapState.mu.Unlock()
		if bootstrapState.managedServer == cmd {
			bootstrapState.managedServerExited = true
			bootstrapState.managedServerExitErr = err
		}
	}(cmd)

	return waitForServer(ctx, resolvedServerAddr, expectedVersion)
}

func shutdownManagedServer() {
	bootstrapState.mu.Lock()
	cmd := bootstrapState.managedServer
	exited := bootstrapState.managedServerExited
	bootstrapState.managedServerBase = ""
	bootstrapState.managedServer = nil
	bootstrapState.managedServerAddr = ""
	bootstrapState.managedServerStdout = nil
	bootstrapState.managedServerStderr = nil
	bootstrapState.managedServerExited = false
	bootstrapState.managedServerExitErr = nil
	bootstrapState.mu.Unlock()

	if cmd != nil && !exited && cmd.Process != nil {
		_ = cmd.Process.Kill()
	}
}

func waitForServer(ctx context.Context, serverAddr string, expectedVersion string) (string, error) {
	deadline := time.Now().Add(20 * time.Second)
	for time.Now().Before(deadline) {
		if details := managedServerFailureDetails(); details != "" {
			shutdownManagedServer()
			return "", fmt.Errorf("allwright server exited before becoming ready at %s\n%s", serverAddr, details)
		}
		if status, err := pingServer(ctx, serverAddr); err == nil && status.version == expectedVersion {
			return serverAddr, nil
		}
		time.Sleep(250 * time.Millisecond)
	}
	details := managedServerFailureDetails()
	shutdownManagedServer()
	if details != "" {
		return "", fmt.Errorf(
			"timed out waiting for allwright server at %s to become ready with version %s\n%s",
			serverAddr,
			expectedVersion,
			details,
		)
	}
	return "", fmt.Errorf("timed out waiting for allwright server at %s to become ready with version %s", serverAddr, expectedVersion)
}

func pingServer(ctx context.Context, serverAddr string) (*pingStatus, error) {
	timeoutCtx, cancel := context.WithTimeout(ctx, time.Second)
	defer cancel()

	conn, err := grpc.NewClient(serverAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return nil, err
	}
	defer conn.Close()

	response, err := enginev1.NewEngineServiceClient(conn).Ping(timeoutCtx, &enginev1.PingRequest{})
	if err != nil {
		return nil, err
	}
	return &pingStatus{version: normalizeReleaseVersion(response.GetVersion())}, nil
}

func ensureCLIAvailable(expectedVersion string) (string, error) {
	if cliPath := strings.TrimSpace(os.Getenv(allwrightCLIPathEnvVar)); cliPath != "" {
		if isFile(cliPath) {
			matches, err := cliVersionMatches(cliPath, expectedVersion)
			if err != nil {
				return "", err
			}
			if matches {
				return cliPath, nil
			}
		}
	}

	home, err := allwrightHome()
	if err != nil {
		return "", err
	}
	bundled := filepath.Join(home, "bin", cliFilename())
	if isFile(bundled) {
		matches, err := cliVersionMatches(bundled, expectedVersion)
		if err != nil {
			return "", err
		}
		if matches {
			return bundled, nil
		}
	}
	if cliPath, ok := repoLocalCLIPath(); ok {
		matches, err := cliVersionMatches(cliPath, expectedVersion)
		if err != nil {
			return "", err
		}
		if matches {
			return cliPath, nil
		}
	}
	if cliPath, err := exec.LookPath(cliFilename()); err == nil {
		matches, err := cliVersionMatches(cliPath, expectedVersion)
		if err != nil {
			return "", err
		}
		if matches {
			return cliPath, nil
		}
	}
	if !autoInstallEnabled() {
		return "", fmt.Errorf("allwright CLI was not found; install it first or set %s", allwrightCLIPathEnvVar)
	}
	return installCLI()
}

func installCLI() (string, error) {
	home, err := allwrightHome()
	if err != nil {
		return "", err
	}
	installDir := filepath.Join(home, "bin")
	if err := os.MkdirAll(installDir, 0o755); err != nil {
		return "", fmt.Errorf("create allwright CLI install dir %s: %w", installDir, err)
	}
	cliPath := filepath.Join(installDir, cliFilename())
	tag, err := resolveReleaseTag()
	if err != nil {
		return "", err
	}
	assetName, err := cliAssetName(tag)
	if err != nil {
		return "", err
	}
	assetBytes, err := downloadReleaseAsset(tag, assetName)
	if err != nil {
		return "", err
	}
	if err := unpackCLIArchive(assetName, assetBytes, cliPath); err != nil {
		return "", err
	}
	return cliPath, nil
}

func ensureWebPlugin(cliPath string, expectedVersion string) error {
	home, err := allwrightHome()
	if err != nil {
		return err
	}
	pluginPath := filepath.Join(home, "plugins", "web", "lib", webPluginFilename())
	if isFile(pluginPath) {
		if version, err := installedPluginVersion(home, "web"); err == nil && version == expectedVersion {
			return nil
		}
	}

	if expectedVersion == "" {
		expectedVersion = normalizeReleaseVersion(defaultReleaseVersion)
	}

	cmd := exec.Command(cliPath, "plugin", "install", "web", "--version", expectedVersion)
	cmd.Stdout = io.Discard
	cmd.Stderr = io.Discard
	if err := cmd.Run(); err != nil {
		return fmt.Errorf("install allwright web plugin: %w", err)
	}
	if !isFile(pluginPath) {
		return fmt.Errorf("allwright attempted to install the web plugin automatically, but the runtime library is still missing")
	}
	if version, err := installedPluginVersion(home, "web"); err == nil && version == expectedVersion {
		return nil
	}
	return fmt.Errorf("allwright attempted to install the web plugin automatically, but version %s is still not active", expectedVersion)
}

func resolveReleaseTag() (string, error) {
	version := strings.TrimSpace(os.Getenv(allwrightVersionEnvVar))
	if version == "" {
		version = defaultReleaseVersion
	}
	if version != "latest" {
		return normalizeReleaseTag(version), nil
	}

	repository := strings.TrimSpace(os.Getenv(allwrightRepositoryEnvVar))
	if repository == "" {
		repository = defaultReleaseRepository
	}
	request, err := http.NewRequest(http.MethodGet, fmt.Sprintf("https://api.github.com/repos/%s/releases/latest", repository), nil)
	if err != nil {
		return "", err
	}
	request.Header.Set("User-Agent", "allwright-go/"+defaultReleaseVersion)
	response, err := http.DefaultClient.Do(request)
	if err != nil {
		return "", err
	}
	defer response.Body.Close()
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return "", fmt.Errorf("resolve latest allwright release: %s", response.Status)
	}
	var payload struct {
		TagName string `json:"tag_name"`
	}
	if err := json.NewDecoder(response.Body).Decode(&payload); err != nil {
		return "", err
	}
	if strings.TrimSpace(payload.TagName) == "" {
		return "", fmt.Errorf("latest allwright release metadata did not include tag_name")
	}
	return payload.TagName, nil
}

func cliAssetName(tag string) (string, error) {
	target := ""
	switch runtime.GOOS + "/" + runtime.GOARCH {
	case "darwin/arm64":
		target = "aarch64-apple-darwin"
	case "darwin/amd64":
		target = "x86_64-apple-darwin"
	case "linux/arm64":
		target = "aarch64-unknown-linux-gnu"
	case "linux/amd64":
		target = "x86_64-unknown-linux-gnu"
	case "windows/arm64":
		target = "aarch64-pc-windows-msvc"
	case "windows/amd64":
		target = "x86_64-pc-windows-msvc"
	default:
		return "", fmt.Errorf("automatic allwright CLI install is not supported on %s/%s", runtime.GOOS, runtime.GOARCH)
	}
	ext := "tar.gz"
	if runtime.GOOS == "windows" {
		ext = "zip"
	}
	return fmt.Sprintf("allwright-%s-%s.%s", tag, target, ext), nil
}

func downloadReleaseAsset(tag, assetName string) ([]byte, error) {
	repository := strings.TrimSpace(os.Getenv(allwrightRepositoryEnvVar))
	if repository == "" {
		repository = defaultReleaseRepository
	}
	request, err := http.NewRequest(http.MethodGet, fmt.Sprintf("https://github.com/%s/releases/download/%s/%s", repository, tag, assetName), nil)
	if err != nil {
		return nil, err
	}
	request.Header.Set("User-Agent", "allwright-go/"+defaultReleaseVersion)
	response, err := http.DefaultClient.Do(request)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return nil, fmt.Errorf("download allwright CLI asset %s: %s", assetName, response.Status)
	}
	return io.ReadAll(response.Body)
}

func unpackCLIArchive(assetName string, assetBytes []byte, destination string) error {
	if strings.HasSuffix(assetName, ".zip") {
		archive, err := zip.NewReader(bytes.NewReader(assetBytes), int64(len(assetBytes)))
		if err != nil {
			return err
		}
		for _, file := range archive.File {
			if file.Name != "bin/"+cliFilename() {
				continue
			}
			reader, err := file.Open()
			if err != nil {
				return err
			}
			defer reader.Close()
			output, err := os.Create(destination)
			if err != nil {
				return err
			}
			defer output.Close()
			if _, err := io.Copy(output, reader); err != nil {
				return err
			}
			return os.Chmod(destination, 0o755)
		}
		return fmt.Errorf("allwright CLI archive did not contain bin/%s", cliFilename())
	}

	gzipReader, err := gzip.NewReader(bytes.NewReader(assetBytes))
	if err != nil {
		return err
	}
	defer gzipReader.Close()
	archive := tar.NewReader(gzipReader)
	for {
		header, err := archive.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			return err
		}
		if header.Name != "bin/"+cliFilename() {
			continue
		}
		output, err := os.Create(destination)
		if err != nil {
			return err
		}
		defer output.Close()
		if _, err := io.Copy(output, archive); err != nil {
			return err
		}
		return os.Chmod(destination, 0o755)
	}
	return fmt.Errorf("allwright CLI archive did not contain bin/%s", cliFilename())
}

func allwrightHome() (string, error) {
	if home := strings.TrimSpace(os.Getenv(allwrightHomeEnvVar)); home != "" {
		return home, nil
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("resolve ALLWRIGHT_HOME: %w", err)
	}
	return filepath.Join(home, ".allwright"), nil
}

func autoInstallEnabled() bool {
	switch strings.ToLower(strings.TrimSpace(os.Getenv(allwrightAutoInstallEnvVar))) {
	case "0", "false", "no":
		return false
	default:
		return true
	}
}

func cliListenAddr(serverAddr string) string {
	return strings.TrimPrefix(strings.TrimPrefix(serverAddr, "http://"), "https://")
}

func normalizeReleaseTag(version string) string {
	version = strings.TrimSpace(version)
	if strings.HasPrefix(version, "v") {
		return version
	}
	return "v" + version
}

func normalizeReleaseVersion(version string) string {
	return strings.TrimPrefix(strings.TrimSpace(version), "v")
}

func expectedRuntimeVersion() string {
	version := strings.TrimSpace(os.Getenv(allwrightVersionEnvVar))
	if version == "" {
		version = defaultReleaseVersion
	}
	return normalizeReleaseVersion(version)
}

func isLocalServerAddr(serverAddr string) bool {
	trimmed := cliListenAddr(serverAddr)
	host := trimmed
	if strings.LastIndex(host, ":") > 0 {
		host = host[:strings.LastIndex(host, ":")]
	}
	host = strings.Trim(host, "[]")
	return host == "127.0.0.1" || host == "localhost" || host == "::1"
}

func installedPluginVersion(home string, pluginID string) (string, error) {
	manifestPath := filepath.Join(home, "plugins.txt")
	contents, err := os.ReadFile(manifestPath)
	if err != nil {
		if os.IsNotExist(err) {
			return "", nil
		}
		return "", err
	}
	for _, line := range strings.Split(string(contents), "\n") {
		trimmed := strings.TrimSpace(line)
		if trimmed == "" || strings.HasPrefix(trimmed, "#") {
			continue
		}
		parts := strings.SplitN(trimmed, "\t", 3)
		if len(parts) < 3 || parts[0] != pluginID {
			continue
		}
		return normalizeReleaseVersion(parts[2]), nil
	}
	return "", nil
}

func allocateManagedServerAddr(serverAddr string) (string, error) {
	host := localBindingHost(serverAddr)
	listener, err := net.Listen("tcp", net.JoinHostPort(host, "0"))
	if err != nil {
		return "", fmt.Errorf("reserve local port for managed allwright server on %s: %w", host, err)
	}
	port := listener.Addr().(*net.TCPAddr).Port
	_ = listener.Close()
	return net.JoinHostPort(host, fmt.Sprintf("%d", port)), nil
}

func repoLocalCLIPath() (string, bool) {
	_, currentFile, _, ok := runtime.Caller(0)
	if !ok {
		return "", false
	}
	repoRoot := filepath.Dir(filepath.Dir(currentFile))
	for _, candidate := range []string{
		filepath.Join(repoRoot, "target", "debug", cliFilename()),
		filepath.Join(repoRoot, "target", "release", cliFilename()),
	} {
		if isFile(candidate) {
			return candidate, true
		}
	}
	return "", false
}

func localBindingHost(serverAddr string) string {
	trimmed := cliListenAddr(serverAddr)
	host := trimmed
	if strings.LastIndex(host, ":") > 0 {
		host = host[:strings.LastIndex(host, ":")]
	}
	host = strings.Trim(host, "[]")
	if host == "::1" {
		return "::1"
	}
	return "127.0.0.1"
}

func displayVersion(version string) string {
	if strings.TrimSpace(version) == "" {
		return "unknown"
	}
	return version
}

func managedServerFailureDetails() string {
	bootstrapState.mu.Lock()
	defer bootstrapState.mu.Unlock()

	if !bootstrapState.managedServerExited && bootstrapState.managedServerExitErr == nil {
		return ""
	}

	parts := make([]string, 0, 3)
	if bootstrapState.managedServerExitErr != nil {
		parts = append(parts, fmt.Sprintf("exit error: %v", bootstrapState.managedServerExitErr))
	}
	if bootstrapState.managedServerStdout != nil {
		if stdout := bootstrapState.managedServerStdout.String(); stdout != "" {
			parts = append(parts, "stdout:\n"+stdout)
		}
	}
	if bootstrapState.managedServerStderr != nil {
		if stderr := bootstrapState.managedServerStderr.String(); stderr != "" {
			parts = append(parts, "stderr:\n"+stderr)
		}
	}
	return strings.Join(parts, "\n")
}

func cliVersionMatches(cliPath string, expectedVersion string) (bool, error) {
	output, err := exec.Command(cliPath, "--version").Output()
	if err != nil {
		return false, fmt.Errorf("inspect allwright CLI version via %s: %w", cliPath, err)
	}
	for _, token := range strings.Fields(string(output)) {
		if token != "" && token[0] >= '0' && token[0] <= '9' {
			return normalizeReleaseVersion(token) == expectedVersion, nil
		}
	}
	return false, nil
}

func cliFilename() string {
	if runtime.GOOS == "windows" {
		return "allwright.exe"
	}
	return "allwright"
}

func webPluginFilename() string {
	switch runtime.GOOS {
	case "darwin":
		return "liballwright_surface_web.dylib"
	case "linux":
		return "liballwright_surface_web.so"
	case "windows":
		return "allwright_surface_web.dll"
	default:
		return "allwright_surface_web.unknown"
	}
}

func isFile(path string) bool {
	info, err := os.Stat(path)
	return err == nil && !info.IsDir()
}

File: python/allwright/_bootstrap.py
from __future__ import annotations

import io
import json
import os
import platform
import shutil
import socket
import subprocess
import tarfile
import tempfile
import threading
import time
import urllib.request
import zipfile
from pathlib import Path

import grpc

from ._proto import DEFAULT_SERVER_ADDR, engine_pb2, engine_pb2_grpc
from ._types import AllwrightError

ALLWRIGHT_AUTO_INSTALL_ENV_VAR = "ALLWRIGHT_AUTO_INSTALL"
ALLWRIGHT_CLI_PATH_ENV_VAR = "ALLWRIGHT_CLI_PATH"
ALLWRIGHT_HOME_ENV_VAR = "ALLWRIGHT_HOME"
ALLWRIGHT_REPOSITORY_ENV_VAR = "ALLWRIGHT_REPOSITORY"
ALLWRIGHT_VERSION_ENV_VAR = "ALLWRIGHT_VERSION"
DEFAULT_RELEASE_REPOSITORY = "allwright-dev/allwright"
DEFAULT_RELEASE_VERSION = "0.0.41"

_bootstrap_lock = threading.Lock()
_managed_server: subprocess.Popen[bytes] | None = None
_managed_server_addr: str | None = None
_managed_server_base_addr: str | None = None


def ensure_runtime_ready(server_addr: str) -> str:
    normalized = (server_addr or DEFAULT_SERVER_ADDR).strip()
    expected_version = expected_runtime_version()
    status = ping_server(normalized)
    if status is not None:
        if status["version"] == expected_version:
            return normalized
        if not is_local_server_addr(normalized):
            raise AllwrightError(
                f"allwright server at {normalized} is running version "
                f"{display_version(status['version'])} but this client expects {expected_version}"
            )
    if not is_local_server_addr(normalized):
        raise AllwrightError(
            f"allwright could not reach engine server at {normalized}. "
            "Automatic startup is only supported for local addresses."
        )

    global _managed_server, _managed_server_addr, _managed_server_base_addr
    with _bootstrap_lock:
        if _managed_server is not None and _managed_server.poll() is None:
            if _managed_server_base_addr == normalized and _managed_server_addr is not None:
                return wait_for_server(_managed_server_addr, expected_version)
            _managed_server.kill()
            _managed_server.wait(timeout=5)
        else:
            _managed_server = None
            _managed_server_addr = None
            _managed_server_base_addr = None

        cli_path = ensure_cli_available(expected_version)
        ensure_web_plugin(cli_path, expected_version)
        resolved_server_addr = normalized
        if status is not None and status["version"] != expected_version:
            resolved_server_addr = allocate_managed_server_addr(normalized)
        _managed_server = subprocess.Popen(
            [str(cli_path), "serve", "--listen-addr", cli_listen_addr(resolved_server_addr)],
            stdin=subprocess.DEVNULL,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        _managed_server_addr = resolved_server_addr
        _managed_server_base_addr = normalized

    return wait_for_server(resolved_server_addr, expected_version)


def shutdown_managed_server() -> None:
    global _managed_server, _managed_server_addr, _managed_server_base_addr
    with _bootstrap_lock:
        if _managed_server is not None and _managed_server.poll() is None:
            _managed_server.kill()
            _managed_server.wait(timeout=5)
        _managed_server = None
        _managed_server_addr = None
        _managed_server_base_addr = None


def wait_for_server(server_addr: str, expected_version: str) -> str:
    deadline = time.time() + 20
    while time.time() < deadline:
        status = ping_server(server_addr)
        if status is not None and status["version"] == expected_version:
            return server_addr
        time.sleep(0.25)
    shutdown_managed_server()
    raise AllwrightError(
        f"timed out waiting for allwright server at {server_addr} "
        f"to become ready with version {expected_version}"
    )


def ping_server(server_addr: str) -> dict[str, str] | None:
    channel = grpc.insecure_channel(server_addr)
    stub = engine_pb2_grpc.EngineServiceStub(channel)
    try:
        response = stub.Ping(engine_pb2.PingRequest(), timeout=1)
        return {"version": normalize_release_version(getattr(response, "version", ""))}
    except grpc.RpcError:
        return None
    finally:
        channel.close()


def ensure_cli_available(expected_version: str) -> Path:
    env_path = os.getenv(ALLWRIGHT_CLI_PATH_ENV_VAR, "").strip()
    if env_path:
        path = Path(env_path)
        if path.is_file() and cli_version_matches(path, expected_version):
            return path

    bundled = allwright_home() / "bin" / cli_filename()
    if bundled.is_file() and cli_version_matches(bundled, expected_version):
        return bundled

    repo_local = repo_local_cli_path()
    if repo_local is not None and repo_local.is_file() and cli_version_matches(repo_local, expected_version):
        return repo_local

    found = shutil.which(cli_filename())
    if found and cli_version_matches(Path(found), expected_version):
        return Path(found)

    if not auto_install_enabled():
        raise AllwrightError(
            "allwright CLI was not found. Install it first or set ALLWRIGHT_CLI_PATH."
        )
    return install_cli()


def install_cli() -> Path:
    install_dir = allwright_home() / "bin"
    install_dir.mkdir(parents=True, exist_ok=True)
    cli_path = install_dir / cli_filename()
    version_tag = resolve_release_tag()
    asset_name = cli_asset_name(version_tag)
    asset_bytes = download_release_asset(version_tag, asset_name)
    unpack_cli_archive(asset_name, asset_bytes, cli_path)
    cli_path.chmod(0o755)
    return cli_path


def ensure_web_plugin(cli_path: Path, expected_version: str) -> None:
    plugin_path = allwright_home() / "plugins" / "web" / "lib" / web_plugin_filename()
    if plugin_path.is_file() and installed_plugin_version("web") == expected_version:
        return

    completed = subprocess.run(
        [str(cli_path), "plugin", "install", "web", "--version", expected_version],
        stdin=subprocess.DEVNULL,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
        check=False,
    )
    if completed.returncode != 0 or not plugin_path.is_file():
        raise AllwrightError(
            "allwright attempted to install the `web` plugin automatically, "
            "but the install did not complete successfully"
        )
    if installed_plugin_version("web") != expected_version:
        raise AllwrightError(
            "allwright attempted to install the `web` plugin automatically, "
            f"but version {expected_version} is still not active"
        )


def resolve_release_tag() -> str:
    version = os.getenv(ALLWRIGHT_VERSION_ENV_VAR, "").strip() or DEFAULT_RELEASE_VERSION
    if version == "latest":
        repository = os.getenv(ALLWRIGHT_REPOSITORY_ENV_VAR, "").strip() or DEFAULT_RELEASE_REPOSITORY
        with urllib.request.urlopen(
            urllib.request.Request(
                f"https://api.github.com/repos/{repository}/releases/latest",
                headers={"User-Agent": f"allwright-python/{DEFAULT_RELEASE_VERSION}"},
            )
        ) as response:
            payload = json.load(response)
        tag = payload.get("tag_name")
        if not isinstance(tag, str) or not tag.strip():
            raise AllwrightError("latest allwright release metadata did not include tag_name")
        return tag
    return normalize_release_tag(version)


def cli_asset_name(version_tag: str) -> str:
    system = platform.system()
    machine = platform.machine()
    normalized_machine = {
        "amd64": "x86_64",
        "arm64": "aarch64",
    }.get(machine.lower(), machine)
    platform_key = (system, normalized_machine)
    targets = {
        ("Darwin", "arm64"): "aarch64-apple-darwin",
        ("Darwin", "aarch64"): "aarch64-apple-darwin",
        ("Darwin", "x86_64"): "x86_64-apple-darwin",
        ("Linux", "aarch64"): "aarch64-unknown-linux-gnu",
        ("Linux", "x86_64"): "x86_64-unknown-linux-gnu",
    }
    target = targets.get(platform_key)
    if target is None and os.name == "nt":
        if normalized_machine == "aarch64":
            target = "aarch64-pc-windows-msvc"
        else:
            target = "x86_64-pc-windows-msvc"
    if target is None:
        raise AllwrightError(
            f"automatic allwright CLI install is not supported on {platform_key[0]}/{platform_key[1]}"
        )
    extension = "zip" if os.name == "nt" else "tar.gz"
    return f"allwright-{version_tag}-{target}.{extension}"


def download_release_asset(version_tag: str, asset_name: str) -> bytes:
    repository = os.getenv(ALLWRIGHT_REPOSITORY_ENV_VAR, "").strip() or DEFAULT_RELEASE_REPOSITORY
    request = urllib.request.Request(
        f"https://github.com/{repository}/releases/download/{version_tag}/{asset_name}",
        headers={"User-Agent": f"allwright-python/{DEFAULT_RELEASE_VERSION}"},
    )
    with urllib.request.urlopen(request) as response:
        return response.read()


def unpack_cli_archive(asset_name: str, asset_bytes: bytes, destination: Path) -> None:
    if asset_name.endswith(".zip"):
        with zipfile.ZipFile(io.BytesIO(asset_bytes)) as archive:
            for member_name in archive.namelist():
                if normalize_archive_path(member_name) != f"bin/{cli_filename()}":
                    continue
                with archive.open(member_name) as source, destination.open("wb") as target:
                    shutil.copyfileobj(source, target)
                return
        raise AllwrightError("allwright CLI archive did not contain the expected binary")
        return

    with tarfile.open(fileobj=io.BytesIO(asset_bytes), mode="r:gz") as archive:
        for member in archive.getmembers():
            if normalize_archive_path(member.name) != f"bin/{cli_filename()}":
                continue
            source = archive.extractfile(member)
            if source is None:
                break
            with source, destination.open("wb") as target:
                shutil.copyfileobj(source, target)
            return
    raise AllwrightError("allwright CLI archive did not contain the expected binary")


def normalize_archive_path(name: str) -> str | None:
    parts: list[str] = []
    for part in Path(name).parts:
        if part in {"", "."}:
            continue
        if part == "..":
            return None
        parts.append(part)
    if not parts:
        return None
    return "/".join(parts)


def allwright_home() -> Path:
    configured = os.getenv(ALLWRIGHT_HOME_ENV_VAR, "").strip()
    if configured:
        return Path(configured)
    return Path.home() / ".allwright"


def auto_install_enabled() -> bool:
    return os.getenv(ALLWRIGHT_AUTO_INSTALL_ENV_VAR, "").strip().lower() not in {
        "0",
        "false",
        "no",
    }


def cli_listen_addr(server_addr: str) -> str:
    if server_addr.startswith("http://"):
        return server_addr[len("http://") :]
    if server_addr.startswith("https://"):
        return server_addr[len("https://") :]
    return server_addr


def normalize_release_tag(version: str) -> str:
    version = version.strip()
    return version if version.startswith("v") else f"v{version}"


def normalize_release_version(version: str) -> str:
    return version.strip().removeprefix("v")


def expected_runtime_version() -> str:
    return normalize_release_version(
        os.getenv(ALLWRIGHT_VERSION_ENV_VAR, "").strip() or DEFAULT_RELEASE_VERSION
    )


def is_local_server_addr(server_addr: str) -> bool:
    host = cli_listen_addr(server_addr).rsplit(":", 1)[0].strip("[]")
    return host in {"127.0.0.1", "localhost", "::1"}


def installed_plugin_version(plugin_id: str) -> str | None:
    manifest = allwright_home() / "plugins.txt"
    if not manifest.is_file():
        return None
    for line in manifest.read_text().splitlines():
        trimmed = line.strip()
        if not trimmed or trimmed.startswith("#"):
            continue
        parts = trimmed.split("\t", 2)
        if len(parts) < 3 or parts[0] != plugin_id:
            continue
        return normalize_release_version(parts[2])
    return None


def allocate_managed_server_addr(server_addr: str) -> str:
    host = local_binding_host(server_addr)
    with socket.create_server((host, 0), family=socket.AF_INET6 if ":" in host else socket.AF_INET) as listener:
        port = listener.getsockname()[1]
    if ":" in host:
        return f"[{host}]:{port}"
    return f"{host}:{port}"


def local_binding_host(server_addr: str) -> str:
    host = cli_listen_addr(server_addr).rsplit(":", 1)[0].strip("[]")
    return "::1" if host == "::1" else "127.0.0.1"


def display_version(version: str) -> str:
    return version or "unknown"


def repo_local_cli_path() -> Path | None:
    repo_root = Path(__file__).resolve().parents[2]
    for candidate in (
        repo_root / "target" / "debug" / cli_filename(),
        repo_root / "target" / "release" / cli_filename(),
    ):
        if candidate.is_file():
            return candidate
    return None


def cli_version_matches(cli_path: Path, expected_version: str) -> bool:
    try:
        completed = subprocess.run(
            [str(cli_path), "--version"],
            stdin=subprocess.DEVNULL,
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
            text=True,
            check=False,
        )
    except OSError as error:
        raise AllwrightError(
            f"inspect allwright CLI version via {cli_path}: {error}"
        ) from error
    if completed.returncode != 0:
        return False
    for token in completed.stdout.split():
        if token and token[0].isdigit():
            return normalize_release_version(token) == expected_version
    return False


def cli_filename() -> str:
    return "allwright.exe" if os.name == "nt" else "allwright"


def web_plugin_filename() -> str:
    if os.name == "nt":
        return "allwright_surface_web.dll"
    if os.uname().sysname == "Darwin":
        return "liballwright_surface_web.dylib"
    return "liballwright_surface_web.so"
